openlibrary/plugins/upstream/addbook.py

```diff
@@ -644,7 +645,12 @@ def save(self, formdata: web.Storage) -> None:
                 edition_data.pop('physical_dimensions', None)
             )
             self.edition.set_weight(edition_data.pop('weight', None))
-            self.edition.set_toc_text(edition_data.pop('table_of_contents', None))
+            try:
+                self.edition.set_toc_text(edition_data.pop('table_of_contents', None))
+            except TocParseError as e:
+                raise ClientException(
+                    "400 Bad Request", f"Table of contents parse error: {e}"
+                )
```

Comment:
Ah, one more little gotcha! We want the user to see the value again so they can fix the error.
```suggestion
            table_of_contents_raw = edition_data.pop('table_of_contents', None)
            try:
                self.edition.set_toc_text(table_of_contents_raw)
            except TocParseError as e:
                # Side-step the parsing and set to the raw value, so the user can
                # keep editing
                self.edition.table_of_contents = table_of_contents_raw
                raise ClientException(
                    "400 Bad Request", f"Table of contents parse error: {e}"
                )
```

Reply:
Errr actually no this won't quite work, we'll need it else where, let me see...